# Setup

In [1]:
%%capture
!pip install --upgrade xgboost

In [2]:
import os
import warnings
warnings.filterwarnings(action='ignore')

import pandas as pd
import numpy as np
pd.set_option('display.max_columns', None)
pd.set_option('precision', 4)
np.set_printoptions(precision=4)

from sklearn.model_selection import KFold
from sklearn.metrics import mean_absolute_error
from xgboost import XGBRegressor

SEED = 23
os.environ['PYTHONHASHSEED'] = str(SEED)
np.random.seed(SEED)

In [3]:
data_url = 'https://raw.githubusercontent.com/stiwari-ds/data-science-competitions/main/dphi/juniper_networks_global_challenge/data/'

train = pd.read_csv(data_url + 'raw/train.csv')
test = pd.read_csv(data_url + 'raw/test.csv')

# Data Overview

In [4]:
train.head()

Unnamed: 0                        observation_id observation_timestamp  \
0           0  704d2a80-d52e-11ec-90ff-c7e6292284b3   2022-05-16 15:39:57   
1           1  1cacc1d0-e6ac-11ec-b65d-156af70ce36b   2022-06-07 21:52:23   
2           2  6dc2b330-d37a-11ec-884e-dfe9ea4a7bd5   2022-05-14 11:38:52   
3           3  163ee0a0-0cca-11ed-a73c-8904b24187cc   2022-07-26 10:02:41   
4           4  5e3c5df0-d5ee-11ec-a5f2-3b6f99e95850   2022-05-17 14:33:50   

   hour_of_day  register__sales_dollar_amt_this_hour  \
0           15                                347.29   
1           21                                361.59   
2           11                                850.73   
3           10                               1175.69   
4           14                               3204.53   

  register__payment_types_accepted  register__peak_sales_dollar_amt_per_hour  \
0                      Cash+Credit                                   -0.7383   
1                      Cash+Credit                                    0.6483   
2                      Cash+Credit                                   -0.4950   
3                      Cash+Credit                                   -0.5594   
4                      Cash+Credit                                    0.5693   

   register__sales_dollar_amt_last_hour  register__sales_quantity_last_hour  \
0                               -0.1270                             -0.1993   
1                               -0.0362                             -0.0777   
2                               -0.1268                             -0.1974   
3                               -0.1270                             -0.1991   
4                               -0.1221                             -0.1632   

   register__sales_quantity_rescanned_frac  \
0                                  -0.8299   
1                                  -0.7395   
2                                   1.3139   
3                                  -0.8299   
4                                  -0.7071   

   register__sales_payments_declined_frac  \
0                                 -0.1247   
1                                 -0.1135   
2                                  0.1075   
3                                 -0.1247   
4                                 -0.1247   

   register__peak_returns_dollar_amt_per_hour  \
0                                     -0.5721   
1                                     -0.5251   
2                                     -0.5251   
3                                     -0.5251   
4                                      0.7604   

   register__returns_dollar_amt_last_hour  \
0                                 -0.2582   
1                                 -0.2157   
2                                 -0.2579   
3                                 -0.2582   
4                                 -0.2565   

   register__returns_quantity_last_hour  \
0                               -0.2621   
1                               -0.1252   
2                               -0.2542   
3                               -0.2616   
4                               -0.2505   

   register__returns_quantity_rescanned_frac  cashier__title_level  \
0                                    -0.5337               -2.6036   
1                                    -0.2853               -0.3520   
2                                    -0.1125               -0.3520   
3                                    -0.2958               -0.3520   
4                                    -0.0756                1.8996   

   cashier__n_years_experience  cashier__hours_into_shift  \
0                      -2.5371                     1.0551   
1                       1.0463                    -0.2628   
2                       1.0463                    -0.6884   
3                       1.0463                    -0.2024   
4                      -0.7454                    -0.3341   

   cashier__item_scan_rate_per_min  cashier__item_manual_entry_rate_per_min  \
0      

In [5]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18512 entries, 0 to 18511
Data columns (total 63 columns):
 #   Column                                       Non-Null Count  Dtype  
---  ------                                       --------------  -----  
 0   Unnamed: 0                                   18512 non-null  int64  
 1   observation_id                               18512 non-null  object 
 2   observation_timestamp                        18512 non-null  object 
 3   hour_of_day                                  18512 non-null  int64  
 4   register__sales_dollar_amt_this_hour         18512 non-null  float64
 5   register__payment_types_accepted             18512 non-null  object 
 6   register__peak_sales_dollar_amt_per_hour     18512 non-null  float64
 7   register__sales_dollar_amt_last_hour         18512 non-null  float64
 8   register__sales_quantity_last_hour           18512 non-null  float64
 9   register__sales_quantity_rescanned_frac      18512 non-null  float64
 10

In [6]:
test.head()

observation_id observation_timestamp  hour_of_day  \
0  3f483640-bc52-11ec-b736-8544dc068949      15/04/2022 00:23            0   
1  a88b0bb0-d2ae-11ec-bd0e-f5a7c7895456      13/05/2022 11:20           11   
2  90611780-d607-11ec-a5d8-675445c92326      17/05/2022 17:34           17   
3  a98380e0-e2d2-11ec-b28f-7f4d381958ee      03/06/2022 00:18            0   
4  608b8e60-e4f2-11ec-8115-d3258731976b      05/06/2022 17:10           17   

  register__payment_types_accepted  register__peak_sales_dollar_amt_per_hour  \
0                      Cash+Credit                                    0.5693   
1                      Cash+Credit                                    0.3084   
2                      Cash+Credit                                   -0.6757   
3                      Cash+Credit                                   -0.1747   
4                      Cash+Credit                                    0.6483   

   register__sales_dollar_amt_last_hour  register__sales_quantity_last_hour  \
0                               -0.1253                             -0.1489   
1                               -0.1270                             -0.1986   
2                                0.1925                              0.2534   
3                               -0.1258                             -0.1709   
4                               -0.0612                             -0.0995   

   register__sales_quantity_rescanned_frac  \
0                                  -0.8176   
1                                   1.6562   
2                                  -0.7517   
3                                  -0.3576   
4                                   0.1060   

   register__sales_payments_declined_frac  \
0                                 -0.1066   
1                                 -0.1247   
2                                 -0.1247   
3                                 -0.1247   
4                                 -0.1247   

   register__peak_returns_dollar_amt_per_hour  \
0                                      0.9103   
1                                     -0.5251   
2                                     -0.5251   
3                                     -0.5251   
4                                     -0.5251   

   register__returns_dollar_amt_last_hour  \
0                                 -0.1930   
1                                 -0.2581   
2                                  0.8384   
3                                 -0.2304   
4                                 -0.2169   

   register__returns_quantity_last_hour  \
0                               -0.2177   
1                               -0.2601   
2                                0.7034   
3                               -0.2174   
4                               -0.1434   

   register__returns_quantity_rescanned_frac  cashier__title_level  \
0                                    -0.5078                1.8996   
1                                    -0.5337               -0.3520   
2                                    -0.2254               -0.3520   
3                                    -0.5308               -0.3520   
4                                    -0.5186               -0.3520   

   cashier__n_years_experience  cashier__hours_into_shift  \
0                      -0.7454                    -0.7090   
1                       1.0463                     0.0167   
2                      -0.7454                    -0.2369   
3                       1.0463                     0.2179   
4                       1.0463                    -0.5244   

   cashier__item_scan_rate_per_min  cashier__item_manual_entry_rate_per_min  \
0                           2.0490                                   1.8238   
1                           1.1217                                   0.9900   
2                           0.0913                                   0.3647   
3                           1.0186                                   0.8858   
4                           0.9156          

In [7]:
test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5179 entries, 0 to 5178
Data columns (total 61 columns):
 #   Column                                       Non-Null Count  Dtype  
---  ------                                       --------------  -----  
 0   observation_id                               5179 non-null   object 
 1   observation_timestamp                        5179 non-null   object 
 2   hour_of_day                                  5179 non-null   int64  
 3   register__payment_types_accepted             5179 non-null   object 
 4   register__peak_sales_dollar_amt_per_hour     5179 non-null   float64
 5   register__sales_dollar_amt_last_hour         5179 non-null   float64
 6   register__sales_quantity_last_hour           5179 non-null   float64
 7   register__sales_quantity_rescanned_frac      5179 non-null   float64
 8   register__sales_payments_declined_frac       5179 non-null   float64
 9   register__peak_returns_dollar_amt_per_hour   5179 non-null   float64
 10  

**Checking missing values**

In [8]:
train.isna().any().any(), test.isna().any().any()

(True, True)

*region__peak_sales_dollar_amt_per_hour_v2* and *region__peak_returns_dollar_amt_per_hour_v2* have missing values in both datasets.  
However, these are the experimental features in the dataset which are not supposed to be used. Thus, we can simply drop these columns instead of imputing the missing values.

**Dropping useless column**

In [9]:
train = train.drop(
    labels=[
        'Unnamed: 0', 
        'region__peak_sales_dollar_amt_per_hour_v2', 
        'region__peak_returns_dollar_amt_per_hour_v2'
    ], 
    axis=1
)

test = test.drop(
    labels=[
        'region__peak_sales_dollar_amt_per_hour_v2', 
        'region__peak_returns_dollar_amt_per_hour_v2'
    ], 
    axis=1
)

# Target

In [10]:
TARGET = 'register__sales_dollar_amt_this_hour'

In [11]:
train[TARGET].min(), train[TARGET].max()

(15.71, 4978.65)

# Features
[Data dictionary](https://docs.google.com/spreadsheets/d/1a_Zw27zQfEZF_GnDqXuoYnoOiyrgOcasw0DIUTAhrr4/edit#gid=1521074975)

### observation_id

In [13]:
train.shape[0], train['observation_id'].nunique()

(18512, 18512)

In [14]:
test.shape[0], test['observation_id'].nunique()

(5179, 5179)

In [15]:
set(train['observation_id'].unique()).intersection(set(test['observation_id'].unique()))

set()

Definitely a unique identifier. Can be dropped from feature set for modeling.

In [16]:
train = train.drop(labels='observation_id', axis=1)
test = test.drop(labels='observation_id', axis=1)

### hour_of_day

In [17]:
train['hour_of_day'].min(), train['hour_of_day'].max()

(0, 23)

In [18]:
test['hour_of_day'].min(), test['hour_of_day'].max()

(0, 23)

### observation_timestamp

In [19]:
train[['observation_timestamp', 'hour_of_day']].sample(5, random_state=SEED)

observation_timestamp  hour_of_day
8125    2022-06-04 07:12:50            7
17419   2022-05-21 20:50:26           20
1859    2022-05-16 21:22:43           21
14474   2022-05-21 01:52:23            1
12401   2022-05-17 20:49:20           20

In [20]:
train['observation_timestamp'] = pd.to_datetime(train['observation_timestamp'])
test['observation_timestamp'] = pd.to_datetime(test['observation_timestamp'])

In [21]:
train.observation_timestamp.min(), train.observation_timestamp.max()

(Timestamp('2022-04-07 21:46:33'), Timestamp('2022-07-29 21:15:48'))

In [22]:
test.observation_timestamp.min(), test.observation_timestamp.max()

(Timestamp('2022-01-06 00:04:00'), Timestamp('2022-12-05 23:56:00'))

In [23]:
len(test.loc[(test.observation_timestamp > '2022-04-07 21:46:33') 
             & (test.observation_timestamp < '2022-07-29 21:15:48')])

3594

* Train dataset records only cover early April to late July 2022, while Test dataset records cover early January to early December 2022.  
We have training data for summer sales but it cannot help predict winter sales.  
* However, 3594 out of 5178 test dataset records fall within our train data range. Thus, there is still some value in the timestamp data.

In [24]:
train['month'] = train.observation_timestamp.dt.month.astype('int')
test['month'] = test.observation_timestamp.dt.month.astype('int')

In [25]:
train['day'] = train.observation_timestamp.dt.day.astype('int')
test['day'] = test.observation_timestamp.dt.day.astype('int')

In [26]:
train['day_of_week'] = train.observation_timestamp.dt.day_of_week.astype('int')
test['day_of_week'] = test.observation_timestamp.dt.day_of_week.astype('int')

### object columns

In [27]:
train.describe(include='O').T

count unique          top   freq
register__payment_types_accepted  18512      3  Cash+Credit  16310
store__type_code                  18512      6            A  11691

#### register__payment_types_accepted

In [28]:
train.register__payment_types_accepted.value_counts(normalize=True)

Cash+Credit          0.8811
Cash+Credit+Check    0.1152
Credit               0.0037
Name: register__payment_types_accepted, dtype: float64

In [29]:
test.register__payment_types_accepted.value_counts(normalize=True)

Cash+Credit          0.8828
Cash+Credit+Check    0.1128
Credit               0.0044
Name: register__payment_types_accepted, dtype: float64

#### store__type_code

In [30]:
train.store__type_code.value_counts(normalize=True)

A    0.6315
C    0.3045
B    0.0465
D    0.0171
F    0.0003
E    0.0001
Name: store__type_code, dtype: float64

In [31]:
test.store__type_code.value_counts(normalize=True)

A    0.6584
C    0.2807
B    0.0423
D    0.0180
F    0.0004
E    0.0002
Name: store__type_code, dtype: float64

#### Encoding:

In [34]:
train['register__payment_types_accepted'] = train['register__payment_types_accepted'].astype('category')
test['register__payment_types_accepted'] = test['register__payment_types_accepted'].astype('category')

train['store__type_code'] = train['store__type_code'].astype('category')
test['store__type_code'] = test['store__type_code'].astype('category')

### numerical columns 

In [35]:
train.describe(include=['float64']).T

count       mean       std  \
register__sales_dollar_amt_this_hour         18512.0  1169.8142  818.4552   
register__peak_sales_dollar_amt_per_hour     18512.0     0.0029    1.0075   
register__sales_dollar_amt_last_hour         18512.0     0.0005    1.1195   
register__sales_quantity_last_hour           18512.0     0.0012    1.1134   
register__sales_quantity_rescanned_frac      18512.0    -0.0144    0.9885   
register__sales_payments_declined_frac       18512.0     0.0015    1.0371   
register__peak_returns_dollar_amt_per_hour   18512.0     0.0067    1.0104   
register__returns_dollar_amt_last_hour       18512.0     0.0031    1.1070   
register__returns_quantity_last_hour         18512.0     0.0026    1.1109   
register__returns_quantity_rescanned_frac    18512.0    -0.0242    0.9607   
cashier__title_level                         18512.0    -0.0018    1.0055   
cashier__n_years_experience                  18512.0    -0.0204    1.0003   
cashier__hours_into_shift                    18512.0     0.0059    0.9928   
cashier__item_scan_rate_per_min              18512.0    -0.0012    1.0071   
cashier__item_manual_entry_rate_per_min      18512.0    -0.0014    1.0089   
store__miles_to_nearest_location             18512.0     0.0300    0.9774   
store__target_sales_quantity_per_hour        18512.0    -0.0227    0.9170   
store__mean_customer_to_staff_ratio          18512.0     0.0007    1.0094   
store__mean_service_time_per_customer        18512.0     0.0161    1.0025   
store__n_employees_total                     18512.0    -0.0118    0.9978   
store__n_managers                            18512.0     0.0025    1.0044   
store__n_baggers                             18512.0     0.0084    1.0092   
store__n_open_registers                      18512.0     0.0141    1.0064   
store__is_sufficiently_staffed               18512.0    -0.0441    0.9500   
store__occupancy_main_floor                  18512.0     0.0028    1.0031   
store__occupancy_grocery                     18512.0     0.0059    0.9975   
store__occupancy_checkout_areas              18512.0    -0.0265    0.9676   
store__occupancy_food_court                  18512.0    -0.0009    0.9950   
store__occupancy_backrooms                   18512.0    -0.0048    0.9944   
store__occupancy_indoors                     18512.0    -0.0024    1.0022   
store__occupancy_outdoors                    18512.0    -0.0215    0.9814   
store__outdoor_temperature                   18512.0    -0.0303    0.9864   
store__parking_lot_utilization               18512.0    -0.0230    0.9969   
store__shelf_freespace_frac                  18512.0    -0.0309    0.9841   
store__hrs_since_last_delivery               18512.0    -0.0101    0.9962   
store__sales_dollar_amt_last_hour            18512.0     0.0024    1.0112   
store__sales_quantity_last_hour              18512.0     0.0027    1.0127   
store__sales_quantity_rescanned_frac         18512.0     0.0005    1.1036   
store__gift_sales_quantity_last_hour         18512.0    -0.0275    0.9752   
store__returns_dollar_amt_last_hour          18512.0    -0.0018    1.0062   
store__returns_quantity_last_hour            18512.0    -0.0003    1.0112   
store__returns_quantity_rescanned_frac       18512.0     0.0069    1.0100   
store__gift_returns_quantity_last_hour       18512.0    -0.0041    0.9717   
region__n_stores                             18512.0     0.0144    1.0105   
region__n_open_registers                     18512.0     0.0174    1.0081   
region__mean_service_time_per_customer       18512.0     0.0210    0.9915   
region__stdev_service_time_per_customer      18512.0     0.0063    0.9975   
region__sales_dollar_amt_last_hour           18512.0     0.0178    1.0155   
region__returns_dollar_amt_last_hour         18512.0     0.0070    1.0396   
region__nighttime_open_registers             18512.0     0.0148    1.0056   
region__nighttime_service_time_per_customer  18512.0     0.0217    0.9788   
region__nighttime_sales_amt_per_hour       

* A lot of negative values where we would not expect them. For instance, median 'sales_dollar_amount_last_hour' is negative. We would expect any kind of sales amount to be non-negative.  
* Except the target variable, the other numerical columns have zero mean and unit std. deviation, which means they are preprocessed (standardized). Thus, we cannot derive any insights from the original data. We can only do further preprocessing such as outlier removal for sensitive algorithms.

# Baseline

In [36]:
features = list(test.columns)
features.remove('observation_timestamp')

In [60]:
def regression_accuracy(y_meas, y_pred, max_error=20, error_type='relative'):
    '''Compares predicted & measured values, returning the percentage of predictions
       that are within a set error limit. This error limit can be an absolute value
       or a relative percentage'''

    # OPTION 1: Relative percentage
    if ( error_type == 'relative' ):
        mask = 100.0 * abs((y_pred - y_meas) / y_meas) < max_error

    # OPTION 2: Absolute value
    if ( error_type == 'absolute' ):
        mask = abs(y_pred - y_meas) < max_error

    accuracy = sum(mask) / len(mask)

    return -accuracy # '-' for xgboost eval-metric usage 

In [61]:
%%time
scores_racc = []
scores_mae = []
test_preds = []

X = train[features]
y = train[TARGET]

baseline = XGBRegressor(
    n_estimators=1000,
    learning_rate = 0.02,
    max_depth=10,
    tree_method='gpu_hist',
    enable_categorical=True,
    max_cat_to_onehot=6,
    predictor='gpu_predictor',
    eval_metric=regression_accuracy,
    early_stopping_rounds=50,
    seed=SEED
)

cv = KFold(n_splits=5, shuffle=True, random_state=SEED)
for fold, (train_idx, val_idx) in enumerate(cv.split(X, y)):
    X_train, y_train = X.loc[train_idx], y.iloc[train_idx]
    X_val, y_val = X.loc[val_idx], y.iloc[val_idx]
    baseline.fit(
        X_train, y_train,
        eval_set=[(X_val, y_val)],
        verbose=50
    )
    preds_val = baseline.predict(X_val)
    racc = regression_accuracy(y_val, preds_val)
    scores_racc.append(racc)
    mae = mean_absolute_error(y_val, preds_val)
    scores_mae.append(mae)
    test_preds.append(baseline.predict(test[features]))
    print(f'Fold #{fold}: R-Acc = {racc:.5f}, MAE = {mae:.5f}')

print(f'\nAvg. R-Acc = {np.mean(scores_racc):.5f} +/- {np.std(scores_racc):.5f}')
print(f'Avg. MAE = {np.mean(scores_mae):.5f} +/- {np.std(scores_mae):.5f}')

[0]	validation_0-rmse:1402.60286	validation_0-regression_accuracy:-0.00000
[50]	validation_0-rmse:548.05768	validation_0-regression_accuracy:-0.05887
[100]	validation_0-rmse:266.91579	validation_0-regression_accuracy:-0.78396
[150]	validation_0-rmse:195.50707	validation_0-regression_accuracy:-0.86335
[200]	validation_0-rmse:180.72904	validation_0-regression_accuracy:-0.86660
[218]	validation_0-rmse:179.51492	validation_0-regression_accuracy:-0.86633
Fold #0: R-Acc = -0.86821, MAE = 112.67949
[0]	validation_0-rmse:1397.52647	validation_0-regression_accuracy:-0.00000
[50]	validation_0-rmse:552.52304	validation_0-regression_accuracy:-0.06940
[100]	validation_0-rmse:284.57429	validation_0-regression_accuracy:-0.78396
[150]	validation_0-rmse:225.57809	validation_0-regression_accuracy:-0.86740
[200]	validation_0-rmse:215.18215	validation_0-regression_accuracy:-0.86740
[226]	validation_0-rmse:213.34882	validation_0-regression_accuracy:-0.86740
Fold #1: R-Acc = -0.87011, MAE = 113.21467
[0]	va

In [64]:
preds_df = pd.DataFrame(np.column_stack(test_preds), columns=[f'Fold_{i}' for i in range(5)])
preds_df

Fold_0     Fold_1     Fold_2     Fold_3     Fold_4
0     1716.9926  1684.6437  1705.6852  1764.9753  1798.0679
1      352.2997   352.0203   351.2715   357.1324   350.3503
2     1368.2958  1375.0820  1361.2714  1410.5143  1376.3406
3      297.1163   290.6902   274.1450   301.2448   279.4444
4      348.9647   353.2632   349.1398   356.0157   354.6789
...         ...        ...        ...        ...        ...
5174  1021.2933  1096.0903  1088.5035  1072.3995  1116.0011
5175  1081.6003  1273.4852  1225.4816  1248.3768  1211.2305
5176   849.8782   839.8718   842.3035   800.8390   816.1514
5177  1076.7432  1084.7090  1245.8694  1016.5651  1061.3091
5178   963.0385   979.0117   951.7444   994.8226   985.0449

[5179 rows x 5 columns]

In [65]:
sub_1 = pd.DataFrame({'prediction': preds_df['Fold_1']})
sub_1.to_csv('01_sub_1.csv', index=False)

In [66]:
sub_0 = pd.DataFrame({'prediction': preds_df['Fold_0']})
sub_0.to_csv('01_sub_0.csv', index=False)

In [69]:
avg_preds = np.mean(np.column_stack(test_preds), axis=1)

In [70]:
sub_avg = pd.DataFrame({'prediction': avg_preds})
sub_avg.to_csv('01_sub_avg.csv', index=False)